include/arch/riscv/arch/32/mode/machine.h

```diff
@@ -9,9 +9,22 @@
 #include <util.h>
 #include <arch/model/smp.h>
 #include <stdint.h>
+#include <plat/machine/devices_gen.h>
+
+#ifdef CONFIG_RISCV_USE_CLINT_MTIME
+/*
+ * Currently all RISC-V 32-bit platforms supported have the mtime register
+ * mapped at the same offset of the base address of the CLINT.
+ */
+#define CLINT_MTIME_OFFSET 0xbff8
+#endif
 
 static inline uint64_t riscv_read_time(void)
 {
+#ifdef CONFIG_RISCV_USE_CLINT_MTIME
+    uint64_t n = *(volatile uint64_t *)(CLINT_PPTR + CLINT_MTIME_OFFSET);
+    return n;
+#else
```

Comment:
It seems like this read can produce invalid values because the compiler breaks it into two 32-bit reads (each `lw` instruction is reading half of the value from the register):
```
static inline uint64_t riscv_read_time(void)
{
#ifdef CONFIG_RISCV_USE_CLINT_MTIME
    uint64_t n = *(volatile uint64_t *)(CLINT_PPTR + CLINT_MTIME_OFFSET);
ff800104:       ffc0c7b7                lui     a5,0xffc0c
ff800108:       ff87a703                lw      a4,-8(a5) # ffc0bff8 <ki_end+0x3f0ff8>
/**
   DONT_TRANSLATE
 */
BOOT_CODE void initTimer(void)
{
    sbi_set_timer(riscv_read_time() + RESET_CYCLES);
ff80010c:       6515                    lui     a0,0x5
ff80010e:       ffc7a783                lw      a5,-4(a5)
```

This probably needs to do a similar thing as the register reads by explicitly requesting a sequence of 32-bit reads and detecting overflow.

Reply:
Good point, I will change that then.

Reply:
I've pushed a commit but there's an issue, see the commit message. I will have to investigate properly when I have more time.

Reply:
https://github.com/seL4/seL4/pull/1076/commits/5cf253ac219e8252d06d7f3ba38dac6ab7e5b33a#r1290876449